File: sdk/typescript/src/phone/resources/contactRules.ts
/**
 * inkbox-phone/resources/contactRules.ts
 *
 * Per-number phone contact rules (allow/block) + org-wide list.
 */

import { HttpTransport } from "../../_http.js";
import { ContactRuleStatus } from "../../mail/types.js";
import {
  PhoneContactRule,
  PhoneRuleAction,
  PhoneRuleMatchType,
  RawPhoneContactRule,
  parsePhoneContactRule,
} from "../types.js";

const ORG_BASE = "/contact-rules";
const NUMBERS_BASE = "/numbers";

function rulePath(phoneNumberId: string, ruleId?: string): string {
  const base = `${NUMBERS_BASE}/${phoneNumberId}/contact-rules`;
  return ruleId ? `${base}/${ruleId}` : base;
}

export interface ListPhoneContactRulesOptions {
  action?: PhoneRuleAction;
  matchType?: PhoneRuleMatchType;
  limit?: number;
  offset?: number;
}

export interface CreatePhoneContactRuleOptions {
  action: PhoneRuleAction;
  matchTarget: string;
  matchType?: PhoneRuleMatchType;
}

export interface UpdatePhoneContactRuleOptions {
  action?: PhoneRuleAction;
  status?: ContactRuleStatus;
}

export interface ListAllPhoneContactRulesOptions {
  phoneNumberId?: string;
  action?: PhoneRuleAction;
  matchType?: PhoneRuleMatchType;
  limit?: number;
  offset?: number;
}

export class PhoneContactRulesResource {
  constructor(private readonly http: HttpTransport) {}

  async list(
    phoneNumberId: string,
    options: ListPhoneContactRulesOptions = {},
  ): Promise<PhoneContactRule[]> {
    const params: Record<string, string | number | undefined> = {};
    if (options.action !== undefined) params.action = options.action;
    if (options.matchType !== undefined) params.match_type = options.matchType;
    if (options.limit !== undefined) params.limit = options.limit;
    if (options.offset !== undefined) params.offset = options.offset;
    const data = await this.http.get<
      { items: RawPhoneContactRule[] } | RawPhoneContactRule[]
    >(rulePath(phoneNumberId), params);
    const items = Array.isArray(data) ? data : data.items;
    return items.map(parsePhoneContactRule);
  }

  async get(phoneNumberId: string, ruleId: string): Promise<PhoneContactRule> {
    const data = await this.http.get<RawPhoneContactRule>(
      rulePath(phoneNumberId, ruleId),
    );
    return parsePhoneContactRule(data);
  }

  async create(
    phoneNumberId: string,
    options: CreatePhoneContactRuleOptions,
  ): Promise<PhoneContactRule> {
    const body: Record<string, unknown> = {
      action: options.action,
      match_type: options.matchType ?? PhoneRuleMatchType.EXACT_NUMBER,
      match_target: options.matchTarget,
    };
    const data = await this.http.post<RawPhoneContactRule>(
      rulePath(phoneNumberId),
      body,
    );
    return parsePhoneContactRule(data);
  }

  async update(
    phoneNumberId: string,
    ruleId: string,
    options: UpdatePhoneContactRuleOptions,
  ): Promise<PhoneContactRule> {
    const body: Record<string, unknown> = {};
    if (options.action !== undefined) body.action = options.action;
    if (options.status !== undefined) body.status = options.status;
    const data = await this.http.patch<RawPhoneContactRule>(
      rulePath(phoneNumberId, ruleId),
      body,
    );
    return parsePhoneContactRule(data);
  }

  async delete(phoneNumberId: string, ruleId: string): Promise<void> {
    await this.http.delete(rulePath(phoneNumberId, ruleId));
  }

  async listAll(
    options: ListAllPhoneContactRulesOptions = {},
  ): Promise<PhoneContactRule[]> {
    const params: Record<string, string | number | undefined> = {};
    if (options.phoneNumberId !== undefined) params.phone_number_id = options.phoneNumberId;
    if (options.action !== undefined) params.action = options.action;
    if (options.matchType !== undefined) params.match_type = options.matchType;
    if (options.limit !== undefined) params.limit = options.limit;
    if (options.offset !== undefined) params.offset = options.offset;
    const data = await this.http.get<
      { items: RawPhoneContactRule[] } | RawPhoneContactRule[]
    >(ORG_BASE, params);
    const items = Array.isArray(data) ? data : data.items;
    return items.map(parsePhoneContactRule);
  }
}

File: sdk/rust/src/phone/resources/contact_rules.rs
//! Phone contact rules (per-number allow/block rules + org-wide list).

use std::sync::Arc;

use serde_json::{Map, Value};

use crate::error::Result;
use crate::http::HttpTransport;
use crate::phone::types::{PhoneContactRule, PhoneRuleAction, PhoneRuleMatchType};

const BASE: &str = "/numbers";
const ORG_BASE: &str = "/contact-rules";

/// Build the per-number contact-rule path, optionally addressing one rule.
fn rule_path(phone_number_id: &str, rule_id: Option<&str>) -> String {
    let base = format!("{BASE}/{phone_number_id}/contact-rules");
    match rule_id {
        None => base,
        Some(id) => format!("{base}/{id}"),
    }
}

/// Unwrap a list response that is either `{"items": [...]}` or a bare array,
/// matching the Python `data["items"] if ... else data` fallback.
fn extract_items(data: Value) -> Result<Vec<PhoneContactRule>> {
    let items = match data {
        Value::Object(mut map) => match map.remove("items") {
            Some(items) => items,
            None => Value::Object(map),
        },
        other => other,
    };
    Ok(serde_json::from_value(items)?)
}

/// Allow/block rules scoped to phone numbers (voice + SMS).
pub struct PhoneContactRulesResource {
    http: Arc<HttpTransport>,
}

impl PhoneContactRulesResource {
    pub fn new(http: Arc<HttpTransport>) -> Self {
        Self { http }
    }

    /// List rules for one phone number.
    pub fn list(
        &self,
        phone_number_id: &str,
        action: Option<PhoneRuleAction>,
        match_type: Option<PhoneRuleMatchType>,
        limit: Option<i64>,
        offset: Option<i64>,
    ) -> Result<Vec<PhoneContactRule>> {
        let mut params: Vec<(&str, String)> = Vec::new();
        if let Some(a) = action {
            params.push(("action", a.as_str().to_string()));
        }
        if let Some(m) = match_type {
            params.push(("match_type", m.as_str().to_string()));
        }
        if let Some(l) = limit {
            params.push(("limit", l.to_string()));
        }
        if let Some(o) = offset {
            params.push(("offset", o.to_string()));
        }
        let data = self.http.get(&rule_path(phone_number_id, None), &params)?;
        extract_items(data)
    }

    /// Get a single rule by ID.
    pub fn get(&self, phone_number_id: &str, rule_id: &str) -> Result<PhoneContactRule> {
        let data = self.http.get(
            &rule_path(phone_number_id, Some(rule_id)),
            crate::http::NO_QUERY,
        )?;
        Ok(serde_json::from_value(data)?)
    }

    /// Create a rule. Use [`update`](Self::update) to change its allow/block
    /// action.
    ///
    /// Returns [`InkboxError::DuplicateContactRule`](crate::error::InkboxError)
    /// on 409 when a non-deleted rule with the same `(match_type, match_target)`
    /// already exists.
    pub fn create(
        &self,
        phone_number_id: &str,
        action: PhoneRuleAction,
        match_target: &str,
        match_type: PhoneRuleMatchType,
    ) -> Result<PhoneContactRule> {
        let mut body = Map::new();
        body.insert("action".into(), action.as_str().into());
        body.insert("match_type".into(), match_type.as_str().into());
        body.insert("match_target".into(), match_target.into());
        let data = self.http.post(
            &rule_path(phone_number_id, None),
            Some(&body),
            crate::http::NO_QUERY,
        )?;
        Ok(serde_json::from_value(data)?)
    }

    /// Update `action` (admin-only).
    pub fn update(
        &self,
        phone_number_id: &str,
        rule_id: &str,
        action: PhoneRuleAction,
    ) -> Result<PhoneContactRule> {
        let mut body = Map::new();
        body.insert("action".into(), action.as_str().into());
        let data = self
            .http
            .patch(&rule_path(phone_number_id, Some(rule_id)), &body)?;
        Ok(serde_json::from_value(data)?)
    }

    /// Delete a rule (admin-only).
    pub fn delete(&self, phone_number_id: &str, rule_id: &str) -> Result<()> {
        self.http.delete(&rule_path(phone_number_id, Some(rule_id)))
    }

    /// Org-wide list of phone contact rules (admin-only).
    ///
    /// # Arguments
    /// * `phone_number_id` - Narrow to a single phone number by id.
    /// * `action` - Filter by `allow` or `block`.
    /// * `match_type` - Filter by `exact_number`.
    pub fn list_all(
        &self,
        phone_number_id: Option<&str>,
        action: Option<PhoneRuleAction>,
        match_type: Option<PhoneRuleMatchType>,
        limit: Option<i64>,
        offset: Option<i64>,
    ) -> Result<Vec<PhoneContactRule>> {
        let mut params: Vec<(&str, String)> = Vec::new();
        if let Some(id) = phone_number_id {
            params.push(("phone_number_id", id.to_string()));
        }
        if let Some(a) = action {
            params.push(("action", a.as_str().to_string()));
        }
        if let Some(m) = match_type {
            params.push(("match_type", m.as_str().to_string()));
        }
        if let Some(l) = limit {
            params.push(("limit", l.to_string()));
        }
        if let Some(o) = offset {
            params.push(("offset", o.to_string()));
        }
        let data = self.http.get(ORG_BASE, &params)?;
        extract_items(data)
    }
}
